### PR TITLE
Setup assistant automating generation of controllers configs

### PR DESCRIPTION
**Pull request by MohmadAyman**
### Description

This PR automates the generation of the ros_controller config files through the Setup Assistant, this is a part of upgrading the setup assistant.  [Issue 894](https://github.com/ros-planning/moveit/issues/894)

This [ros_controllers.yaml](https://gist.github.com/MohmadAyman/0e02336956cdec3b270d0e49f386b6b5) was produced using this PR foe the PR2 Robot. 

I'll follow this PR with another one that exposes the options through a GUI, and add documentation for it.

### Checklist
- [ ] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Include a screenshot if changing a GUI
- [ ] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 1]
```diff
@@ -284,12 +284,11 @@ bool ConfigurationFilesWidget::loadGenFiles()
   file.write_on_changes = 0;  // Can they be changed?
   gen_files_.push_back(file);
 
-  // fake_controllers.yaml --------------------------------------------------------------------------------------
-  file.file_name_ = "fake_controllers.yaml";
+  // controllers.yaml --------------------------------------------------------------------------------------
+  file.file_name_ = "controllers.yaml";
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: ros_controller.yaml

**Comment by MohmadAyman** [comment]
https://gist.github.com/MohmadAyman/76abce5525170ec6e0b119787c44490f

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 2]
```diff
@@ -251,7 +251,7 @@ class MoveItConfigData
   bool outputOMPLPlanningYAML(const std::string& file_path);
   bool outputKinematicsYAML(const std::string& file_path);
   bool outputJointLimitsYAML(const std::string& file_path);
-  bool outputFakeControllersYAML(const std::string& file_path);
```
Comment (davetcoleman) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: lets keep the fake controllers also

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/controllers.launch [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
```
Comment (davetcoleman) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/controllers.launch]: ros_controllers.launch

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 4]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
 }
 
+// ******************************************************************************************
+// Output controllers config files
+// ******************************************************************************************
+bool MoveItConfigData::outputControllersYAML(const std::string& file_path)
+{
+  // Cache the joints' names.
+  std::vector<std::vector<std::string>> groups;
+  std::vector<std::string> groupJoints;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: please use underscore for var names per ros style guidelines

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 5]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
   return true;  // file created successfully
 }
 
+// ******************************************************************************************
+// Output controllers config files
+// ******************************************************************************************
+bool MoveItConfigData::outputControllersYAML(const std::string& file_path)
+{
+  // Cache the joints' names.
+  std::vector<std::vector<std::string>> groups;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: use more specific name than just groups

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 6]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+  std::vector<std::string> groupJoints;
+
+  // Loop through groups
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: expand comment: why loop through?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 7]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+    // Get list of associated joints
+    const robot_model::JointModelGroup* joint_model_group = getRobotModel()->getJointModelGroup(group_it->name_);
+    const std::vector<const robot_model::JointModel*>& joint_models = joint_model_group->getActiveJointModels();
+    // Iterate through the joints
+    for (const robot_model::JointModel* joint : joint_models)
+    {
+      if (joint->isPassive() || joint->getMimic() != NULL || joint->getType() == robot_model::JointModel::FIXED)
+        continue;
+      else
+      {
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: don't mix using brackets without using brackets... all or none

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 8]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+  std::vector<std::string> groupJoints;
+
+  // Loop through groups
+  for (std::vector<srdf::Model::Group>::iterator group_it = srdf_->groups_.begin(); group_it != srdf_->groups_.end();
+       ++group_it)
+  {
+    // Get list of associated joints
+    const robot_model::JointModelGroup* joint_model_group = getRobotModel()->getJointModelGroup(group_it->name_);
+    const std::vector<const robot_model::JointModel*>& joint_models = joint_model_group->getActiveJointModels();
+    // Iterate through the joints
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: explain why you are iterating through the joints

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 9]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+  emitter << YAML::Comment("Settings for ros_control hardware interface");
+  emitter << YAML::BeginMap;
+  emitter << YAML::Key << "hardware_interface";
+  emitter << YAML::Value << YAML::BeginMap;
+
+  emitter << YAML::Key << "groups";
+  emitter << YAML::Value << YAML::BeginSeq;
+
+  // Loop through groups
+  for (int i = 0; i < srdf_->groups_.size(); ++i)
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: std::size_t

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 10]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+  {
+    emitter << YAML::BeginMap;
+    emitter << YAML::Key << "name";
+    emitter << YAML::Value << srdf_->groups_[i].name_ + "_controller";
+    emitter << YAML::Key << "joints";
+    emitter << YAML::Value << YAML::BeginSeq;
+
+    // Iterate through the joints
+    for (int j = 0; j < groups[i].size(); j++)
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: std::size_t

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 11]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+  emitter << YAML::Comment("0: position, 1: velocity");
+
+  emitter << YAML::EndMap;
+
+  emitter << YAML::Newline;
+  emitter << YAML::Comment("Publish all joint states");
+  emitter << YAML::Newline << YAML::Comment("Creates the /joint_states topic necessary in ROS");
+  emitter << YAML::EndMap;
+
+  emitter << YAML::BeginMap;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: what if in order to track the begin and end of maps, brackets were used to indent these sections? e.g.

```
emitter << YAML::BeginMap;
{
    emitter << YAML::Key << "joint_state_controller";
    emitter << YAML::Value << YAML::BeginMap;
    ...
}
emitter << YAML::EndMap;
```

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 12]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+  std::ofstream output_stream(file_path.c_str(), std::ios_base::trunc);
+  if (!output_stream.good())
+  {
+    ROS_ERROR_STREAM("Unable to open file for writing " << file_path);
+    return false;
+  }
+  output_stream << emitter.c_str();
+  output_stream.close();
+
+  return true;  // file created successfully
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: this function is very long and readability is low. can it be separated out into a few smaller functions for each section of the yaml?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 13]
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
...
+  }
+
+  emitter << YAML::EndSeq;
+  emitter << YAML::EndMap;
+  emitter << YAML::EndMap;
+
+  std::ofstream output_stream(file_path.c_str(), std::ios_base::trunc);
+  if (!output_stream.good())
+  {
+    ROS_ERROR_STREAM("Unable to open file for writing " << file_path);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: use ROS_ERROR_STREAM_NAMED

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch [hunk 14]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<launch>
+
+  <!-- Load joint controller configurations from YAML file to parameter server -->
+  <rosparam file="$(find [GENERATED_PACKAGE_NAME])/config/controllers.yaml" command="load"/>
+
+  <!-- load the controllers -->
```
Comment (davetcoleman) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: be consistent with capitalization in comments. i personally prefer capitalizing

**Comment by davetcoleman** [comment]
@mcevoyandy please review

**Comment by MohmadAyman** [comment]
Sorry didn't see the comments till now, I'll have a look at them later today.
For now I got the output configuration to work with '[moveit_sim_controller](https://github.com/davetcoleman/moveit_sim_controller)'.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 13 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Why do you recommend using the Named version here?
All the functions here that cant output a file uses this
`ROS_ERROR_STREAM("Unable to open file for writing " << file_path);`

**Comment by MohmadAyman** [comment]
Addressed the comments, please have a look. 
this is a [ros_conttollers.yaml](https://gist.github.com/MohmadAyman/0e02336956cdec3b270d0e49f386b6b5) produced for the ur5 and tested with moveit_sim_controller
@davetcoleman  @mcevoyandy

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch [hunk 15]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<launch>
+
+  <!-- Load joint controller configurations from YAML file to parameter server -->
+  <rosparam file="$(find [GENERATED_PACKAGE_NAME])/config/ros_controllers.yaml" command="load"/>
+
+  <!-- Load the controllers -->
+  <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
+    output="screen" ns="/[GENERATED_PACKAGE_NAME]" args="spawn joint_state_controller position_trajectory_controller"/>
```
Comment (rhaschke) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: @willcbaker Is using absolute namespaces compatible to 4598f6602ded91d91fccabe7a8bcc490a528fc54?

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+  <!-- Convert joint states to TF transforms for rviz, etc -->
+  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
+    respawn="false" output="screen">
+    <remap from="/joint_states" to="/[GENERATED_PACKAGE_NAME]/joint_states" />
```
Comment (rhaschke) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: Same here.

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 17]
```diff
@@ -254,6 +254,7 @@ class MoveItConfigData
   bool outputKinematicsYAML(const std::string& file_path);
   bool outputJointLimitsYAML(const std::string& file_path);
   bool outputFakeControllersYAML(const std::string& file_path);
+  bool outputControllersYAML(const std::string& file_path);
```
Comment (rhaschke) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: outputROSControllersYAML?

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch (hunk 15 above)
Comment (willcbaker) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: I don't believe so, I would advise against absolute namespacing.
Unable to test or review until next week, seems changing this way with generated package name will require pushing down into package name namespace?

Comment (MohmadAyman) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: So you recommend to remove the `ns` argument from here and from the `robot_state_publisher` node? right?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 12 above)
Comment (davetcoleman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Sure, no problem.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 13 above)
Comment (davetcoleman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: See the style guidelines under "ROS":
http://moveit.ros.org/documentation/contributing/code/

It makes filtering all the command line output much easier and cleaner. All of the MSA should be updated to use NAMED, but that belongs in a separate simple PR that doesn't clutter this one.

Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Okay I understand, Thanks.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 18]
```diff
@@ -732,7 +732,7 @@ bool MoveItConfigData::outputControllersYAML(const std::string& file_path)
   std::ofstream output_stream(file_path.c_str(), std::ios_base::trunc);
   if (!output_stream.good())
   {
-    ROS_ERROR_STREAM("Unable to open file for writing " << file_path);
+    ROS_ERROR_STREAM_NAMED("ros_controllers_config", "Unable to open file for writing " << file_path);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: how about namespace = "setup_assistant"

Comment (rhaschke) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: -1 The package name (which is moveit_setup_assistant) is already part of the logger name. No need for an additional name at all!

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch [hunk 19]
```diff
@@ -6,12 +6,12 @@
 
   <!-- Load the controllers -->
   <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
-    output="screen" ns="/[GENERATED_PACKAGE_NAME]" args="spawn joint_state_controller position_trajectory_controller"/>
+    output="screen" ns="[ROBOT_NAME]" args="spawn joint_state_controller position_trajectory_controller"/>
```
Comment (rhaschke) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: Looks like the ns prefix is very specific to your application. Why do you prefer `ROBOT_NAME` over `GENERATED_PACKAGE_NAME` now?

Comment (MohmadAyman) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: After looking at [gazebo_ros_demos/rrbot_control.launch](https://github.com/ros-simulation/gazebo_ros_demos/blob/kinetic-devel/rrbot_control/launch/rrbot_control.launch) , I've found we use the robot name not the package name for the namespace.
Which is also done in the [ros_control tutorial here](http://gazebosim.org/tutorials/?tut=ros_control].

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 18 above)
Comment (davetcoleman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Good point about the duplicate namespace, I should have recommended the file name "moveit_config_data" as that's the standard we usually follow. I really do like using _NAMED though. I've just attempted to document this standard better:
https://github.com/ros-planning/moveit.ros.org/pull/166

Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: So do we get back to `ROS_ERROR_STREAM`?

Comment (rhaschke) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: That's my suggestion. @davetcoleman?

Comment (davetcoleman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: I prefer the consistency of always using NAMED as I've just attempted to document in https://github.com/ros-planning/moveit.ros.org/pull/166. Perhaps a Github issue should be opened to specifically debate this. Either way, having the namespace certainly does not hurt anything so I would like to just merge this PR.

Comment (rhaschke) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: @davetcoleman I don't agree. If you just have a redundant logger name like xxx.xxx, there is no reason to insist in `NAMED`. When we recently reworked the logging in some other MoveIt packages, we also dropped the extra NAME in such cases.

**Comment by rhaschke** [comment]
@MohmadAyman Please do not merge the devel branch into pull-request branches - never ever.
Instead you should rebase your PR to the latest devel branch. 
Otherwise the commit history is screwed up and we have a really hard time to follow in the git tree what happened. Also, we cannot easily squash-merge the PR.

I manually merged this now in 89f55aaa7721351ce59cb1c0c09bd2b4141988c2.

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch (hunk 15 above)
Comment (rhaschke) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: Yes, I think so.

Comment (MohmadAyman) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch]: This `ros_controllers.launch` is intended to be used with `gazebo`, and according to [`ros_control` tutorials](http://gazebosim.org/tutorials/?tut=ros_control#Addthegazebo_ros_controlplugin), this is how `gazebo_ros_control` plugin is defined in the `urdf`
```
<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
    <robotNamespace>/MYROBOT</robotNamespace>
  </plugin>
</gazebo>
```
So if we were to launch the controllers with not the `robot_name` as a `namespace` , and leave the plugin element `<robotNamespace>`  blank, how would we be able to use simulate multiple robot for example?
Is there a solution for a situation like that?

**Pull request closed without merging**